Extend using statement (#12121)
ethereum/solidity: docs/contracts/using-for.rst

These functions will receive the object they are called on
as their first parameter (like the ``self`` variable in Python).

It is valid either at file level or inside a contract,
at contract level.

The first part, ``A``, can be one of:

- a list of file-level or library functions (``using {f, g, h, L.t} for uint;``) -
  only those functions will be attached to the type.
- the name of a library (``using L for uint;``) -
  all functions (both public and internal ones) of the library are attached to the type

At file level, the second part, ``B``, has to be an explicit type (without data location specifier).
Inside contracts, you can also use ``using L for *;``,
which has the effect that all functions of the library ``L``
are attached to *all* types.

If you specify a library, *all* functions in the library are attached,
even those where the type of the first parameter does not
match the type of the object. The type is checked at the
point the function is called and function overload
resolution is performed.

If you use a list of functions (``using {f, g, h, L.t} for uint;``),
then the type (``uint``) has to be implicitly convertible to the

> first parameter of each of these functions. This mean the check is

Reviewer: ```suggestion
first parameter of each of these functions. This check is
```
or otherwise ``means``

performed even if none of these functions are called.

The ``using A for B;`` directive is active only within the current
scope (either the contract or the current module/source unit),
including within all of its functions, and has no effect
outside of the contract or module in which it is used.

Let us rewrite the set example from the

> :ref:`libraries` in this way, using file-level functions

Reviewer: ```suggestion
:ref:`libraries` section in this way, using file-level functions
```
Or otherwise the ``the`` before is weird, isn't it?

instead of library functions.

.. code-block:: solidity

    // SPDX-License-Identifier: GPL-3.0
    pragma solidity ^0.8.12;

    struct Data { mapping(uint => bool) flags; }
    // Now we attach functions to the type.

>     // This is valid in the rest of the module.

Reviewer: ```suggestion
    // The attached functions can be used throughout the rest of the module.
```

    // If you import the module, you have to

>     // repeat the statement there, potentially as

Reviewer: ```suggestion
    // repeat the using statement there, for example as
```

        }
    }

> It is also possible to extend elementary types in that way.

Reviewer: Strictly speaking, the array below is not an "elementary type", though, is it?
Not sure we need to change anything, though.
